File: src/main/java/io/github/toberocat/guiengine/xml/GuiComponentDeserializer.java
package io.github.toberocat.guiengine.xml;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonNode;
import io.github.toberocat.guiengine.GuiEngineApi;
import io.github.toberocat.guiengine.components.GuiComponent;
import io.github.toberocat.guiengine.components.GuiComponentBuilder;
import io.github.toberocat.guiengine.context.GuiContext;
import io.github.toberocat.guiengine.utils.ParserContext;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.util.UUID;

/**
 * Custom JSON deserializer for a specific type of {@link GuiComponent}.
 * This class is responsible for deserializing JSON data into a {@link GuiComponent} object using its builder.
 *
 * @param <C> The type of {@link GuiComponent} to be deserialized.
 * @param <B> The type of {@link GuiComponentBuilder} associated with the component to be deserialized.
 *            <p>
 *            Created: 10.07.2023
 *            Author: Tobias Madlberger (Tobias)
 */
public class GuiComponentDeserializer<C extends GuiComponent, B extends GuiComponentBuilder> extends JsonDeserializer<C> {

    private final @NotNull Class<B> builderClazz;

    /**
     * Constructor for the GuiComponentDeserializer.
     *
     * @param builderClazz The class of the {@link GuiComponentBuilder} associated with the component to be deserialized.
     */
    public GuiComponentDeserializer(@NotNull Class<B> builderClazz) {
        this.builderClazz = builderClazz;
    }

    /**
     * Deserialize JSON data into a {@link GuiComponent} object using its builder.
     *
     * @param p                      The {@link JsonParser} to read the JSON data from.
     * @param deserializationContext The {@link DeserializationContext} to use during deserialization.
     * @return A {@link GuiComponent} object created from the JSON data using its builder.
     * @throws IOException If an I/O error occurs during JSON parsing.
     */
    @Override
    public C deserialize(JsonParser p, DeserializationContext deserializationContext) throws IOException {
        B builder;
        try {
            builder = builderClazz.getConstructor().newInstance();
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException |
                 NoSuchMethodException e) {
            throw new RuntimeException("Builder " + builderClazz.getName() + " has no default constructor (0 arguments)", e);
        }

        JsonNode node = p.getCodec().readTree(p);

        String apiId = node.get("__:api:__").asText();
        UUID contextId = UUID.fromString(node.get("__:ctx:__").asText());

        GuiEngineApi api = GuiEngineApi.APIS.get(apiId);
        GuiContext context = GuiEngineApi.LOADED_CONTEXTS.get(contextId);

        builder.deserialize(new ParserContext(node, context, api));
        return (C) builder.createComponent();
    }
}

File: src/main/kotlin/io/github/toberocat/guiengine/xml/GuiComponentDeserializer.kt
package io.github.toberocat.guiengine.xml

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.JsonNode
import io.github.toberocat.guiengine.GuiEngineApi
import io.github.toberocat.guiengine.components.GuiComponent
import io.github.toberocat.guiengine.components.GuiComponentBuilder
import io.github.toberocat.guiengine.exception.GuiException
import io.github.toberocat.guiengine.xml.parsing.ParserContext
import java.io.IOException
import java.lang.reflect.InvocationTargetException
import java.util.*

/**
 * Custom JSON deserializer for a specific type of [GuiComponent].
 * This class is responsible for deserializing JSON data into a [GuiComponent] object using its builder.
 *
 * @param <C> The type of [GuiComponent] to be deserialized.
 * @param <B> The type of [GuiComponentBuilder] associated with the component to be deserialized.
 *
 *
 * Created: 10.07.2023
 * Author: Tobias Madlberger (Tobias)
</B></C> */
class GuiComponentDeserializer<C : GuiComponent, B : GuiComponentBuilder?>(private val builderClazz: Class<B>) :
    JsonDeserializer<C>() {

    /**
     * Deserialize JSON data into a [GuiComponent] object using its builder.
     *
     * @param p                      The [JsonParser] to read the JSON data from.
     * @param deserializationContext The [DeserializationContext] to use during deserialization.
     * @return A [GuiComponent] object created from the JSON data using its builder.
     * @throws IOException If an I/O error occurs during JSON parsing.
     */
    @Throws(IOException::class)
    override fun deserialize(p: JsonParser, deserializationContext: DeserializationContext): C {
        val builder: B = try {
            builderClazz.getConstructor().newInstance()
        } catch (ex: Exception) {
            when (ex) {
                is InstantiationException, is IllegalAccessException, is InvocationTargetException, is NoSuchMethodException ->
                    throw RuntimeException("Builder ${builderClazz.name} has no default constructor (0 arguments)", ex)

                else -> throw ex
            }
        }

        val node = p.codec.readTree<JsonNode>(p)
        val apiId = node["__:api:__"].asText()
        val contextId = UUID.fromString(node["__:ctx:__"].asText())
        val api = GuiEngineApi.APIS[apiId]
            ?: throw GuiException("Couldn't parse component. Interpreter didn't specify '__:api:__' to a correct api id")
        val context = GuiEngineApi.LOADED_CONTEXTS[contextId]
            ?: throw GuiException("Couldn't parse component. Value received: $contextId. Interpreter didn't specify '__:ctx:__' to a correct context id")

        val computables: MutableMap<String, String> = mutableMapOf()
        builder!!.deserialize(ParserContext(node, computables, context, api))

        val component = builder.createComponent() as C
        computables.forEach { (field, value) ->
            context.computableFunctionProcessor.markAsComputed(
                component,
                field,
                value
            )
        }

        return component
    }
}
